framework/include/mfem/utils/CoefficientMap.h

```diff
@@ -0,0 +1,189 @@
+                          this->_iterable_coefficients.cend(),
+                          coeff) != this->_iterable_coefficients.cend(),
+                "Coefficient object was not created by this CoefficientMap.");
+
+    const auto [_, inserted] = this->_coefficients.emplace(name, std::move(coeff));
+    if (!inserted)
+    {
+      throw MooseException("Coefficient with name '" + name +
+                           "' already present in CoefficientMap object");
+    }
+  }
+
+  /// Add piecewise material property. The coefficient must have been created with the `make` method on this object.
+  ///
+  /// Note: If you attempt to overwrite an existing block then an exception will be thrown and data
+  /// for that property will be left in an undefined state.
+  void addPiecewiseBlocks(const std::string & name,
+                          std::shared_ptr<T> coeff,
+                          const std::vector<std::string> & blocks)
+  {
+    // If list of blocks is empty then treat as a global coefficient
+    if (blocks.size() == 0)
+    {
+      this->addCoefficient(name, coeff);
+      return;
+    }
+
+    // Initialise property with empty coefficients, if it does not already exist
+    if (!this->hasCoefficient(name))
+    {
+      this->_coefficients.insert({name, this->emptyPWData(coeff)});
+    }
+    PWData * data = std::get_if<PWData>(&this->_coefficients[name]);
+    // Throw an exception if the data is not piecewise
+    if (!data)
+    {
+      throw MooseException("Global coefficient with name '" + name +
+                           "' already present in CoefficientMap");
+    }
+    mooseAssert(std::find(this->_iterable_coefficients.cbegin(),
+                          this->_iterable_coefficients.cend(),
+                          coeff) != this->_iterable_coefficients.cend(),
+                "Coefficient object was not created by the appropriate coefficient manager.");
+    auto & [pw_coeff, coeff_map] = *data;
+    this->checkPWData(coeff, pw_coeff, name);
+
+    for (const auto & block : blocks)
+    {
+      if (coeff_map.count(block) > 0)
+      {
+        throw MooseException("Property with name '" + name + "' already assigned to block " +
+                             block + " in CoefficientMap object");
+      }
+      coeff_map[block] = coeff;
+      pw_coeff->UpdateCoefficient(std::stoi(block), *coeff);
+    }
+  }
+
+  T & getCoefficient(const std::string & name) { return *this->getCoefficientPtr(name); }
+
+  std::shared_ptr<T> getCoefficientPtr(const std::string & name)
+  {
+    try
+    {
+      auto & coeff = this->_coefficients.at(name);
+      try
+      {
+        return std::get<std::shared_ptr<T>>(coeff);
+      }
+      catch (const std::bad_variant_access &)
+      {
+        return std::get<0>(std::get<PWData>(coeff));
+      }
+    }
+    catch (const std::out_of_range &)
+    {
+      throw MooseException("Property with name '" + name + "' has not been declared.");
```

Comment:
Are these recoverable exceptions? Typically we reserve `MooseException` for 'transient' situations that come up during solves due to the values in the solution vector which may change if we, say, cut the time step. Otherwise we typically use `mooseError`

Reply:
We change `mooseError` to throw for unit tests 😄